### DISCUSSION TITLE
More flexible relative time ranges.

### DISCUSSION THREAD

**Discussion by jeme** [grafana/grafana#71028] [Dashboards, Panels & Visualizations]
Currently the relative time ranges only support "now-{value}[wdhms]"

In all cases where I have previously implemented such features, and subsequently worked with them we have supported a more flexible format of where examples are:

```
-5h30m               // now is implicit, so = now - 5 hours and 30 minutes
now-5hours 10seconds // it's ok to skip a unit and units can be spelled out fully
now-1hour 30min      // units does not have to be in plural, short form for minutes and seconds supported.
now-1h 30m 10sec     // spaces are allowed between units for better readability
today-10d12h30m      // today can be used to start the relative range at today at 00:00.
+1d30h30m            // positive ranges are also allowed
1day                 // positive ranges are the default so + is optional
```

Because of this I just naturally tried to input now-12h30min into grafana which obviously failed.

This is also supported by having a combined Datetime + Dateoffset picker where users can switch between offseets or absolute where automatic calculations then take place.
![image](https://github.com/grafana/grafana/assets/1609615/3fc3e96c-2300-4016-9d36-28ed0e0e3a98)
(We have not needed seconds as a granularity, although the underlying parser supports down to ms actually, we have also not added weeks but both are things that could be taken into account)

-> So when we go from Date Offset Picker to Date Time Picker, the offsets are converted to absolute date 
times, and then back if we do the reverse.
![image](https://github.com/grafana/grafana/assets/1609615/9ff56c20-5359-4b1a-9c8d-5e99a86878b2)

To begin with, adding support for an extended format for relative ranges (datetime offsets) seems to be fairly easy, if we exclude today and positive ranges (they don't seem to make much sense for grafana, at least not for the primary usecase, but I could be wrong) it looks like it's a simple matter of replacing a regex and a method in `grafana/packages/grafana-ui/src/components/DateTimePickers/RelativeTimeRangePicker/utils.ts` + adjustments to the reverse method.

Such format could be a simplified version: `now?-({weeks}w)?({days}d)?({hours}h)?({minutes}m)?({seconds}s)?` or go all in.

Then obviously adding a huge bunch of tests. Adding an enhanced UI may be a bit more work, especially considering that the current picker is an actual Date range picker so that you pick both from and to in a single UI, which becomes really clunky in design if time has to be added on top, which is why we often opt for having two individual pickers instead.

I would not mind trying to contribute at least parts of this.

It would be nice if there was a guide to get grafana running in a codespace for that of if anyone had tried that and could give a few hints, but that is not required.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). We are migrating this discussion to an issue and closing the discussion. The issue is #82597. Feel free to continue the discussion around this there. Thank you!
